File: game/app/editor.rb
require_relative 'scene'

class EditorScene < Scene
  RENDER_TARGET_ID = :editor_scene
  DISPLAY_WIDTH = 960
  DISPLAY_HEIGHT = 540
  SCALE = DISPLAY_WIDTH / SCREEN_WIDTH

  def initialize(id)
    super(id)
    @font = Font.new(:font, 16)
  end

  def draw
    $args.outputs[RENDER_TARGET_ID].transient!
    $args.outputs[RENDER_TARGET_ID].background_color = [0, 0, 0, 255]
    $args.outputs[RENDER_TARGET_ID].w = SCREEN_WIDTH
    $args.outputs[RENDER_TARGET_ID].h = SCREEN_HEIGHT

    (0...TILES_X).each do |i|
      (0...TILES_Y).each do |j|
        next unless wall?(i, j)
        @tileset[@tiles[i][j]].draw(i * TILE_SIZE, j * TILE_SIZE, render_target_id: RENDER_TARGET_ID)
      end
    end
    (0..TILES_X).each do |i|
      (0..TILES_Y).each do |j|
        tl = i == 0 || j == 0 || @tiles[i - 1][j - 1]
        tr = i == TILES_X || j == 0 || @tiles[i][j - 1]
        bl = i == 0 || j == TILES_Y || @tiles[i - 1][j]
        br = i == TILES_X || j == TILES_Y || @tiles[i][j]
        next unless tl && tr && bl && br
        Window.draw_rect((i - 0.5) * TILE_SIZE, (j - 0.5) * TILE_SIZE, TILE_SIZE, TILE_SIZE, 0xff000000, render_target_id: RENDER_TARGET_ID)
      end
    end
    (0...TILES_X).each do |i|
      (0...TILES_Y).each do |j|
        Window.draw_rect(i * TILE_SIZE, j * TILE_SIZE, TILE_SIZE, TILE_SIZE, 0x50ffff00, render_target_id: RENDER_TARGET_ID) if wall?(i, j)
        Window.draw_rect(i * TILE_SIZE, j * TILE_SIZE, TILE_SIZE, TILE_SIZE, 0x5000ff00, render_target_id: RENDER_TARGET_ID) if @tiles[i][j] == -1
      end
    end

    @entrances.each_with_index do |(col, row), index|
      x = col * TILE_SIZE
      y = row * TILE_SIZE
      Window.draw_rect(x, y, TILE_SIZE, TILE_SIZE, 0xff0000ff, render_target_id: RENDER_TARGET_ID)
      @font.draw_text(index.to_s, SCALE * x, SCALE * y, 0xffffffff)
    end
    @exits.each do |e|
      x = e.col * TILE_SIZE
      y = e.row * TILE_SIZE
      Window.draw_rect(x, y, TILE_SIZE, TILE_SIZE, 0xffff0000, render_target_id: RENDER_TARGET_ID)
      @font.draw_text("#{e.dest_scene},#{e.dest_entrance}", SCALE * x, SCALE * y, 0xffffffff)
    end

    @objects.each { |o| o.draw(render_target_id: RENDER_TARGET_ID) }

    $args.outputs.primitives << {
      path: RENDER_TARGET_ID,
      x: 0,
      y: Window.height - DISPLAY_HEIGHT,
      w: DISPLAY_WIDTH,
      h: DISPLAY_HEIGHT,
      source_w: SCREEN_WIDTH,
      source_h: SCREEN_HEIGHT
    }
  end

  def add_wall(col, row)
    return if existing_object?(col, row)

    @tiles[col][row] = 0
    check_wall_tiles(col, row)
  end

  def add_wall_edge(col, row)
    return if existing_object?(col, row)

    @tiles[col][row] = -1
    check_wall_tiles(col, row)
  end

  def add_entrance(col, row)
    return if @tiles[col][row]
    return if existing_object?(col, row)

    @entrances << [col, row]
  end

  def add_exit(col, row, dest_scene, dest_entr)
    return if @tiles[col][row]
    return if existing_object?(col, row)

    @exits << Exit.new(col, row, dest_scene, dest_entr)
  end

  def add_object(col, row, class_name, args)
    return if @tiles[col][row]
    return if existing_object?(col, row)

    @objects << Object.const_get(class_name).new(col, row, args.split(","))
  end

  def delete_at(col, row)
    if @tiles[col][row]
      @tiles[col][row] = nil
      check_wall_tiles(col, row)
      return
    end

    @entrances.reject! { |(i, j)| i == col && j == row }
    @exits.reject! { |x| x.col == col && x.row == row }
    @objects.reject! { |x| x.col == col && x.row == row }
  end

  def serialize
    lines = [
      @entrances.map { |e| e.join(',') }.join('|'),
      @exits.map { |e| "#{e.col},#{e.row},#{e.dest_scene},#{e.dest_entrance}" }.join('|'),
      @objects.map { |o| "#{o.class},#{o.col},#{o.row}#{o.args ? ",#{o.args.join(',')}" : ''}" }.join('|'),
    ]
    lines += @tiles.transpose.map { |row| row.map { |tile| tile_char_from_number(tile) }.join }
    lines.join("\n")
  end

  private

  def check_wall_tiles(col, row)
    set_wall_tile(col, row)
    set_wall_tile(col, row - 1) if row > 0
    set_wall_tile(col + 1, row) if col < TILES_X - 1
    set_wall_tile(col, row + 1) if row < TILES_Y - 1
    set_wall_tile(col - 1, row) if col > 0
  end

  def existing_object?(col, row)
    return true if @entrances.find { |(i, j)| i == col && j == row }
    return true if @exits.find { |x| x.col == col && x.row == row }
    @objects.find { |x| x.col == col && x.row == row }
  end

  def tile_char_from_number(tile)
    return '_' if tile.nil?
    return '#' if tile >= 0
    '/'
  end
end

class Editor
  OBJECT_FILES_TO_EXCLUDE = %w[base_object index trigger].freeze

  class << self
    def init
      @scene = EditorScene.new(1)
      @exit_dest_scene = 1
      @exit_dest_entr = 0

      @object_names = ($gtk.list_files("app/object").map { |s| s.chomp(".rb") } - OBJECT_FILES_TO_EXCLUDE).map(&:capitalize)
      @object_index = 0

      @level_id = 1

      font = Font.new(:font, 32)
      @controls = [
        (@lbl_tool = Label.new(EditorScene::DISPLAY_WIDTH + 10, 10, font, "[none]", color: 0xffffff)),
        Button.new(10, 10, w: 40, h: 40, anchor: :top_right, font: font, text: '#') { set_tool(:wall) },
        Button.new(10, 60, w: 40, h: 40, anchor: :top_right, font: font, text: '/') { set_tool(:wall_edge) },
        Button.new(10, 110, w: 40, h: 40, anchor: :top_right, font: font, text: 'e') { set_tool(:entrance) },
        Button.new(10, 160, w: 40, h: 40, anchor: :top_right, font: font, text: 'x') { set_tool(:exit) },
        (lbl_dest_scene = Label.new(50, 210, font, '1', anchor: :top_right, color: 0xffffff)),
        Button.new(10, 210, w: 32, h: 32, anchor: :top_right, font: font, text: '>') { @exit_dest_scene += 1; lbl_dest_scene.text = @exit_dest_scene.to_s },
        Button.new(90, 210, w: 32, h: 32, anchor: :top_right, font: font, text: '<') { if @exit_dest_scene > 1; @exit_dest_scene -= 1; lbl_dest_scene.text = @exit_dest_scene.to_s; end },
        (lbl_dest_entr = Label.new(50, 252, font, '0', anchor: :top_right, color: 0xffffff)),
        Button.new(10, 252, w: 32, h: 32, anchor: :top_right, font: font, text: '>') { @exit_dest_entr += 1; lbl_dest_entr.text = @exit_dest_entr.to_s },
        Button.new(90, 252, w: 32, h: 32, anchor: :top_right, font: font, text: '<') { if @exit_dest_entr > 0; @exit_dest_entr -= 1; lbl_dest_entr.text = @exit_dest_entr.to_s; end },
        Button.new(10, 294, w: 40, h: 40, anchor: :top_right, font: font, text: 'o') { set_tool(:object) },
        (lbl_obj_name = Label.new(50, 344, font, @object_names[0], anchor: :top_right, color: 0xffffff)),
        Button.new(10, 344, w: 32, h: 32, anchor: :top_right, font: font, text: '>') do
          @object_index = (@object_index + 1) % @object_names.size
          lbl_obj_name.text = @object_names[@object_index]
          @txt_obj_args.text = ""
        end,
        Button.new(180, 344, w: 32, h: 32, anchor: :top_right, font: font, text: '<') do
          @object_index = (@object_index - 1) % @object_names.size
          lbl_obj_name.text = @object_names[@object_index]
          @txt_obj_args.text = ""
        end,
        (@txt_obj_args = TextField.new(10, 386, anchor: :top_right, font: font, h: 32)),
        (lbl_level = Label.new(50, 428, font, '1', anchor: :top_right, color: 0xffffff)),
        Button.new(10, 428, w: 32, h: 32, anchor: :top_right, font: font, text: '>') { @level_id += 1; lbl_level.text = @level_id.to_s },
        Button.new(90, 428, w: 32, h: 32, anchor: :top_right, font: font, text: '<') { if @level_id > 1; @level_id -= 1; lbl_level.text = @level_id.to_s; end },
        Button.new(10, 470, h: 40, anchor: :top_right, font: font, text: 'Save') { $gtk.write_file("data/scene/#{@level_id}.txt", @scene.serialize) },
        Button.new(10, 520, h: 40, anchor: :top_right, font: font, text: 'Load') { @scene = EditorScene.new(@level_id) },
      ]
    end

    def update
      @controls.each(&:update)
      return unless Mouse.over?(0, 0, EditorScene::DISPLAY_WIDTH, EditorScene::DISPLAY_HEIGHT)

      col = (Mouse.x / (TILE_SIZE * EditorScene::SCALE)).floor
      row = (Mouse.y / (TILE_SIZE * EditorScene::SCALE)).floor
      if Mouse.button_down?(:left)
        case @active_tool
        when :wall
          @scene.add_wall(col, row)
        when :wall_edge
          @scene.add_wall_edge(col, row)
        when :entrance
          @scene.add_entrance(col, row)
        when :exit
          @scene.add_exit(col, row, @exit_dest_scene, @exit_dest_entr)
        when :object
          @scene.add_object(col, row, @object_names[@object_index], @txt_obj_args.text)
        end
      elsif Mouse.button_down?(:right)
        @scene.delete_at(col, row)
      end
    end

    def draw
      $args.outputs.solids << { x: 0, y: 0, w: EDITOR_SCREEN_WIDTH, h: EDITOR_SCREEN_HEIGHT, r: 0, g: 0, b: 0, a: 255 }
      @scene.draw
      @controls.each(&:draw)

      (1..TILES_X).each { |i| Window.draw_rect(i * TILE_SIZE * EditorScene::SCALE, 0, 1, EditorScene::DISPLAY_HEIGHT, 0x80ffffff) }
      (1..TILES_Y).each { |i| Window.draw_rect(0, i * TILE_SIZE * EditorScene::SCALE, EditorScene::DISPLAY_WIDTH, 1, 0x80ffffff) }
    end

    private

    def set_tool(tool)
      @active_tool = tool
      @lbl_tool.text = "[#{tool}]"
    end
  end
end

File: src/editor.lua
require("src.constants")
require("src.scene")

SCENE_DISPLAY_WIDTH = 960
SCENE_DISPLAY_HEIGHT = 540
SCENE_SCALE = SCENE_DISPLAY_WIDTH / SCREEN_WIDTH

EditorScene = setmetatable({}, Scene)
EditorScene.__index = EditorScene

function EditorScene.new(id)
  local self = Scene.new(id, true)
  setmetatable(self, EditorScene)
  self.canvas = love.graphics.newCanvas(SCREEN_WIDTH, SCREEN_HEIGHT)
  return self
end

function EditorScene:draw()
  love.graphics.setCanvas(self.canvas)
  love.graphics.clear(BG_COLOR)

  Scene.draw(self)

  for i = 1, TILES_X do
    for j = 1, TILES_Y do
      if self:is_wall(i, j) then Window.draw_rectangle((i - 1) * TILE_SIZE, (j - 1) * TILE_SIZE, TILE_SIZE, TILE_SIZE, {1, 1, 0, 0.2}) end
      if self.tiles[i][j] == -1 then Window.draw_rectangle((i - 1) * TILE_SIZE, (j - 1) * TILE_SIZE, TILE_SIZE, TILE_SIZE, {0, 1, 0, 0.2}) end
    end
  end

  love.graphics.setCanvas()
  love.graphics.draw(self.canvas, 0, 0, nil, SCENE_SCALE, SCENE_SCALE)

  for index, e in ipairs(self.entrances) do
    local x = (e[1] - 1) * SCENE_SCALE * TILE_SIZE
    local y = (e[2] - 1) * SCENE_SCALE * TILE_SIZE
    Window.draw_rectangle(x, y, SCENE_SCALE * TILE_SIZE, SCENE_SCALE * TILE_SIZE, {0, 0, 1})
    Editor.font:draw_text(index, x + 2, y + 2)
  end
  for index, e in ipairs(self.exits) do
    local x = (e.col - 1) * SCENE_SCALE * TILE_SIZE
    local y = (e.row - 1) * SCENE_SCALE * TILE_SIZE
    Window.draw_rectangle(x, y, SCENE_SCALE * TILE_SIZE, SCENE_SCALE * TILE_SIZE, {1, 0, 0})
    Editor.font:draw_text(e.dest_scene .. "," .. e.dest_entrance, x + 2, y + 2)
  end
end

function EditorScene:add_wall(col, row)
  if self:existing_object(col, row) then return end

  self.tiles[col][row] = 0
  self:check_wall_tiles(col, row)
end

function EditorScene:add_wall_edge(col, row)
  self.tiles[col][row] = -1
  self:check_wall_tiles(col, row)
end

function EditorScene:add_entrance(col, row)
  if self.tiles[col][row] or self:existing_object(col, row) then return end

  table.insert(self.entrances, {col, row})
end

function EditorScene:add_exit(col, row, dest_scene, dest_entr)
  if self.tiles[col][row] or self:existing_object(col, row) then return end

  table.insert(self.exits, Exit.new(col, row, dest_scene, dest_entr))
end

function EditorScene:add_object(col, row, class_name, args)
  if self:is_wall(col, row) or self:existing_object(col, row) then return end

  local obj = _G[class_name].new(col, row, Utils.split(args, ","))
  obj.class_name = class_name
  table.insert(self.objects, obj)
end

function EditorScene:delete_at(col, row)
  if self.tiles[col][row] then
    self.tiles[col][row] = nil
    self:check_wall_tiles(col, row)
    return
  end

  for i, e in ipairs(self.entrances) do
    if e[1] == col and e[2] == row then
      table.remove(self.entrances, i)
      break
    end
  end
  for i, e in ipairs(self.exits) do
    if e.col == col and e.row == row then
      table.remove(self.exits, i)
      break
    end
  end
  for i, o in ipairs(self.objects) do
    if o.col == col and o.row == row then
      table.remove(self.objects, i)
      break
    end
  end
end

function EditorScene:serialize()
  local lines = {
    Utils.join(Utils.map(self.entrances, function(e) return e[1] .. "," .. e[2] end), "|"),
    Utils.join(Utils.map(self.exits, function(e) return e.col .. "," .. e.row .. "," .. e.dest_scene .. "," .. e.dest_entrance end), "|"),
    Utils.join(Utils.map(self.objects, function(o) return o.class_name .. "," .. o.col .. "," .. o.row .. "," .. (o.args and Utils.join(o.args, ",") or "") end), "|"),
  }
  for j = 1, TILES_Y do
    local line = ""
    for i = 1, TILES_X do
      line = line .. self:tile_char_from_number(i, j)
    end
    table.insert(lines, line)
  end
  return Utils.join(lines, "\n")
end

function EditorScene:check_wall_tiles(col, row)
  self:set_wall_tile(col, row)
  if row > 1 then self:set_wall_tile(col, row - 1) end
  if col < TILES_X then self:set_wall_tile(col + 1, row) end
  if row < TILES_Y then self:set_wall_tile(col, row + 1) end
  if col > 1 then self:set_wall_tile(col - 1, row) end
end

function EditorScene:existing_object(col, row)
  for i, e in ipairs(self.entrances) do
    if e[1] == col and e[2] == row then return true end
  end
  for i, e in ipairs(self.exits) do
    if e.col == col and e.row == row then return true end
  end
  for i, o in ipairs(self.objects) do
    if o.col == col and o.row == row then return true end
  end
end

function EditorScene:tile_char_from_number(i, j)
  local tile = self.tiles[i][j]
  if tile == nil then return "_" end
  if tile >= 0 then return "#" end
  return "/"
end

Editor = {
  init = function ()
    Window.set_size(false, WINDOW_WIDTH, WINDOW_HEIGHT, WINDOW_WIDTH, WINDOW_HEIGHT)
    if not love.filesystem.getInfo("editor") then
      love.filesystem.createDirectory("editor")
    end

    Editor.scene = EditorScene.new(1)
    Editor.exit_dest_scene = 1
    Editor.exit_dest_entr = 1

    Editor.object_names = Utils.map(love.filesystem.getDirectoryItems("src/object"), function(full_name)
      local name = string.gsub(table.remove(Utils.split(full_name, "/")), ".lua", "")
      return Utils.join(Utils.map(Utils.split(name, "_"), function(s) return s:sub(1, 1):upper() .. s:sub(2, -1) end))
    end)
    for i = 1, #Editor.object_names do
      if Editor.object_names[i] == "BaseObject" or Editor.object_names[i] == "Index" then
        table.remove(Editor.object_names, i)
        i = i - 1
      end
    end
    Editor.object_index = 1
    Editor.level_id = 1

    local font = Res.img_font("font", "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzÁÉÍÓÚÀÃÕÂÊÔÑÇáéíóúàãõâêôñç0123456789.,:;!?¡¿/\\()[]+-%'\"←→∞$#<>", 1)
    Editor.lbl_tool = Label.new(SCENE_DISPLAY_WIDTH + 10, 10, {font = font, text = "[none]", scale = 2})
    Editor.lbl_dest_scene = Label.new(50, 214, {font = font, text = "1", anchor = "top_right", scale = 2})
    Editor.lbl_dest_entr = Label.new(50, 256, {font = font, text = "1", anchor = "top_right", scale = 2})
    Editor.lbl_obj_name = Label.new(50, 348, {font = font, text = Editor.object_names[1], anchor = "top_right", scale = 2})
    Editor.txt_obj_args = TextField.new(10, 386, {anchor = "top_right", font = font, h = 32, scale = 2})
    Editor.lbl_level = Label.new(50, 432, {font = font, text = "1", anchor = "top_right", scale = 2})

    Editor.controls = {
      Editor.lbl_tool,
      Button.new(10, 10, {w = 40, h = 40, anchor = "top_right", font = font, text = '#', scale = 2}, function () Editor.set_tool("wall") end),
      Button.new(10, 60, {w = 40, h = 40, anchor = "top_right", font = font, text = '/', scale = 2}, function () Editor.set_tool("wall_edge") end),
      Button.new(10, 110, {w = 40, h = 40, anchor = "top_right", font = font, text = 'e', scale = 2}, function () Editor.set_tool("entrance") end),
      Button.new(10, 160, {w = 40, h = 40, anchor = "top_right", font = font, text = 'x', scale = 2}, function () Editor.set_tool("exit") end),
      Editor.lbl_dest_scene,
      Button.new(10, 210, {w = 32, h = 32, anchor = "top_right", font = font, text = '>', scale = 2}, function ()
        Editor.exit_dest_scene = Editor.exit_dest_scene + 1
        Editor.lbl_dest_scene:set_text(tostring(Editor.exit_dest_scene))
      end),
      Button.new(90, 210, {w = 32, h = 32, anchor = "top_right", font = font, text = '<', scale = 2}, function ()
        if Editor.exit_dest_scene > 1 then
          Editor.exit_dest_scene = Editor.exit_dest_scene - 1
          Editor.lbl_dest_scene:set_text(tostring(Editor.exit_dest_scene))
        end
      end),
      Editor.lbl_dest_entr,
      Button.new(10, 252, {w = 32, h = 32, anchor = "top_right", font = font, text = '>', scale = 2}, function ()
        Editor.exit_dest_entr = Editor.exit_dest_entr + 1
        Editor.lbl_dest_entr:set_text(tostring(Editor.exit_dest_entr))
      end),
      Button.new(90, 252, {w = 32, h = 32, anchor = "top_right", font = font, text = '<', scale = 2}, function ()
        if Editor.exit_dest_entr > 1 then
          Editor.exit_dest_entr = Editor.exit_dest_entr - 1
          Editor.lbl_dest_entr:set_text(tostring(Editor.exit_dest_entr))
        end
      end),
      Button.new(10, 294, {w = 40, h = 40, anchor = "top_right", font = font, text = 'o', scale = 2}, function () Editor.set_tool("object") end),
      Editor.lbl_obj_name,
      Button.new(10, 344, {w = 32, h = 32, anchor = "top_right", font = font, text = '>', scale = 2}, function ()
        Editor.object_index = Editor.object_index % #Editor.object_names + 1
        Editor.lbl_obj_name:set_text(Editor.object_names[Editor.object_index])
        Editor.txt_obj_args:set_text("")
      end),
      Button.new(180, 344, {w = 32, h = 32, anchor = "top_right", font = font, text = '<', scale = 2}, function ()
        Editor.object_index = Editor.object_index == 1 and #Editor.object_names or Editor.object_index - 1
        Editor.lbl_obj_name:set_text(Editor.object_names[Editor.object_index])
        Editor.txt_obj_args:set_text("")
      end),
      Editor.txt_obj_args,
      Editor.lbl_level,
      Button.new(10, 428, {w = 32, h = 32, anchor = "top_right", font = font, text = '>', scale = 2}, function ()
        Editor.level_id = Editor.level_id + 1
        Editor.lbl_level:set_text(tostring(Editor.level_id))
      end),
      Button.new(90, 428, {w = 32, h = 32, anchor = "top_right", font = font, text = '<', scale = 2}, function ()
        if Editor.level_id > 1 then
          Editor.level_id = Editor.level_id - 1
          Editor.lbl_level:set_text(tostring(Editor.level_id))
        end
      end),
      Button.new(10, 470, {h = 40, anchor = "top_right", font = font, text = 'Save', scale = 2}, function ()
        love.filesystem.write("editor/" .. Editor.level_id .. ".txt", Editor.scene:serialize())
      end),
      Button.new(10, 520, {h = 40, anchor = "top_right", font = font, text = 'Load', scale = 2}, function ()
        Editor.scene = EditorScene.new(Editor.level_id)
      end),
    }
    Editor.font = font
  end,
  update = function ()
    KB.update()
    Mouse.update()
    for _, c in ipairs(Editor.controls) do c:update() end
    if not Mouse.over(0, 0, SCENE_DISPLAY_WIDTH, SCENE_DISPLAY_HEIGHT) then
      Editor.mouse_pos = nil
      return
    end

    local col = math.floor(Mouse.x / (TILE_SIZE * SCENE_SCALE)) + 1
    local row = math.floor(Mouse.y / (TILE_SIZE * SCENE_SCALE)) + 1
    if Mouse.down("left") then
      if Editor.active_tool == "wall" then
        Editor.scene:add_wall(col, row)
      elseif Editor.active_tool == "wall_edge" then
        Editor.scene:add_wall_edge(col, row)
      elseif Editor.active_tool == "entrance" then
        Editor.scene:add_entrance(col, row)
      elseif Editor.active_tool == "exit" then
        Editor.scene:add_exit(col, row, Editor.exit_dest_scene, Editor.exit_dest_entr)
      elseif Editor.active_tool == "object" then
        Editor.scene:add_object(col, row, Editor.object_names[Editor.object_index], Editor.txt_obj_args.text)
      end
    elseif Mouse.down("right") then
      Editor.scene:delete_at(col, row)
    end

    Editor.mouse_pos = {col, row}
  end,
  draw = function ()
    love.graphics.clear(0, 0, 0)
    Editor.scene:draw()
    for _, c in ipairs(Editor.controls) do c:draw() end

    for i = 1, TILES_X - 1 do
      Window.draw_rectangle(i * TILE_SIZE * SCENE_SCALE, 0, 1, SCENE_DISPLAY_HEIGHT, {0, 0, 0, 0.5})
    end
    for i = 1, TILES_Y - 1 do
      Window.draw_rectangle(0, i * TILE_SIZE * SCENE_SCALE, SCENE_DISPLAY_WIDTH, 1, {0, 0, 0, 0.5})
    end

    if Editor.mouse_pos then
      local text = Editor.mouse_pos[1] .. "," .. Editor.mouse_pos[2]
      Editor.font:draw_text(text, Mouse.x - Editor.font:text_width(text), Mouse.y - Editor.font.height)
    end
  end,
  set_tool = function(tool)
    Editor.active_tool = tool
    Editor.lbl_tool:set_text("[" .. tool .. "]")
  end
}
